### ISSUE TITLE
Proposal: Filesystem integrity check

### ISSUE THREAD

**Issue by sul3n3t** [docker/docker#32153]
## Proposal: Filesystem integrity check
Create a build option which records a digest of an image's read-only filesystem. It would be stored in the image manifest. At container run-time, if a filesystem digest is present in the manifest, the filesystem is verified against the digest before container execution. If verification fails, an error would be thrown and container execution stopped.

The filesystem digest must be stable, computing the same value on different systems (though on the same platform). The current consideration is [continuity](https://github.com/stevvooe/continuity). The digest would be included in the image manifest, as an optional field of `RootFS`.

This proposal only pertains to filesystem integrity. Other portions of the data or distribution pipeline are out-of-scope.

## Background
Docker images have some mechanisms for ensuring data integrity - that image data at run-time is the same as data at build-time. Docker notary can supply signed digests for an image manifest. The image manifest is downloaded as _image@digest_ and the digest verified at download.

Image manifests contain a `RootFS` JSON object with a sequence of digests of layers. Layers are tarballs that are downloaded, uncompressed, and metadata for tar-split persisted. The layer digest is checked at download. If all layers were downloaded (and digests verified) for new containers and restarts, the container filesystem should be authentic and match data from build-time. However, layer data is cached locally and the original tarball discarded, so digest checks at download do not guarantee current data integrity.

A property of (read-only) layers is the original tar stream can be re-created, and optionally checked against the expected digest. However, verifying the tar stream is not the same as the container filesystem. There may be additional files or files masked by whiteouts, which do not appear in the tar stream.

### dm-verity
[dm-verity](https://gitlab.com/cryptsetup/cryptsetup/wikis/DMVerity) is an excellent way to ensure integrity of a read-only block device. It is not well suited for container images. A container filesystem is composed of stacked layers of decompressed tarballs. The result is not block-equivalent to a build system, but logically equivalent as files. Additionally, loopback devices are necessary for each dm-verify mount. There are resource limits on the number of loopback devices which would limit the number of containers or layers, depending on the implementation.

### Downloading tarball every time or persisting original tarball
If the original, digest verified data was available, it would be possible to re-created a filesystem from verified data. However, the additional storage space, network availability, bandwidth, and latency necessary would be onerous.

### Verifying re-created tar stream of layers
As discussed in [background](#background), tar stream data can be re-created from layers. This initially seems promising for ensuring the integrity of file data. However, the tar stream is not exactly the same as the filesystem data. When recreating the tar stream, only select files are inspected. Any additional files would not be inspected. Files masked by whiteouts may appear in the tar stream, but not in the filesystem. The is also a breaking of abstraction by comparing layer data to the filesystem. The filesystem is prepared by the configured graphdriver. Graphdrivers have very little high-level data and the layer store has no information on how layers are overlayed by the graphdriver.

### Computing digest on first install
Because layer data is verified at download, it is possible to compute a digest of the filesystem at first use. For later uses of the filesystem, the digest from first use can be calculated again and compared. With this approach, the computed digest is stored so only root on the host OS may modify the digest file. This is the same level of security of the cached layer data itself. Additionally, there is no way to audit the computed digest against a known good value.

**Comment by thaJeztah** [comment]
ping @tonistiigi @stevvooe ptal

**Comment by stevvooe** [comment]
All in all, this seems like a reasonable approach.

I think a kind of "half-way" approach would be to generate a continuity manifest _directly_ from the tar file at unpack time. When the layer is unpacked, you can then verify against this manifest. This would avoid the need to supporting a new `RootFS` parameter and distributing continuity manifests.

**Comment by tonistiigi** [comment]
I would prefer the "half-way" if possible. The manifest generation doesn't even need to be on unpack. It can just run in the background and just needs to be completed before the container is created. For the source of this manifest generation, you would use the metadata from tar-split, and extracted files from disk. While reading the file data from the disk you would verify that the layerID by tarstream is still correct.

Adding extra value to `RootFS` is a bit inconsistent with the intermediate images generated during build because this value would only be added to the final image. It would need to be behind a flag. In general, it would be nice to just replace the tar hash with continuity hash if that has better properties, but I don't see a good upgrade path for that.

**Comment by sul3n3t** [comment]
The problem with the half-way approach is that is can only be done when a layer is unpacked from the tarball. Integrity of unpacked layers on disk is definitely a goal here.

**Comment by stevvooe** [comment]
> The problem with the "half-way" approach is that is can only be done when a layer is unpacked from the tarball. Integrity of unpacked layers on disk is definitely a goal here.

I'm not sure that is true. You can keep the manifest around and re-verify to your hearts content. For layers that are _already_ unpacked when the feature is enabled, this won't be possible, but it is reasonable to wipe the machine before enabling this feature.

**Comment by sul3n3t** [comment]
The option I'm proposing would be a build flag. Only new images, which opt-in to using the flag would build a filesystem digest.

There is already some integrity protection of individual layers. For example, the hash of the tarball is useful for a downloader to verify a binary asset. The reproducible tar stream from a layer is useful for the integrity of files from tar stream, independent of any other layers. However, there is no protection from **extra** files on disk not from the tar stream or files masked by another layer.

**Comment by tonistiigi** [comment]
@sul3n3t SGTM. Can't think of a better solution that wouldn't involve multiple passes of retarring everything on every container startup. To detect extra files you would generate continuity manifest directly from tar-split and compare it with the one from the filesystem.

Do we have someone using continuity atm? How do we verify that it is solid and doesn't need breaking updates before we start putting these hashes in the images.

**Comment by stevvooe** [comment]
> However, there is no protection from extra files on disk not from the tar stream or files masked by another layer.

It sounds like we want to mount the filesystem and check it against the continuity manifest.

> Do we have someone using continuity atm? How do we verify that it is solid and doesn't need breaking updates before we start putting these hashes in the images.

It is growing in interest. There was a fairly extensive test about a year ago but I'd like to avoid stabilizing the format as a hash referenced item (although, "round trip" on a manifest is a non-goal for continuity).

**Comment by endophage** [comment]
That's why it's experimental for now. We can also add a version identifier such that future updates can choose how to handle deprecated versions of the integrity check.

**Comment by sul3n3t** [comment]
I forgot to write that this would be experimental in the initial proposal. Added now.

**Comment by AkihiroSuda** [comment]
@sul3n3t

> The option I'm proposing would be a build flag. Only new images, which opt-in to using the flag would build a filesystem digest.

Just a question.
Does this proposal add the sha256 digest of continutiy manifest to the image JSON, and push it to the registry, right?
(If not, how to verify the continutiy manifest against an image digest, which would be specified as `docker run foo@sha256:deadbeef`?)

Also, I think we need to define some "mask manifest" so as to ignore some properties of continutiy manifest, such as [beagle xattrs](https://www.freedesktop.org/wiki/CommonExtendedAttributes/).

**Comment by sul3n3t** [comment]
@AkihiroSuda That's correct, the continuity manifest digest would be included in the image config JSON. So, if the image config JSON has integrity, the integrity of the applied layers can follow.

There is definitely room for masking. For example, a little masking/processing occurs for working with user namespaces.
